builder/etcd/etcd.go

```diff
@@ -0,0 +1,529 @@
+// - path (string): The path to watch
+//
+// Returns:
+// 	- chan bool: Send this a message to stop the watcher.
+func Watch(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	// etcdctl -C $ETCD watch --recursive /deis/services
+	path := p.Get("path", "/deis/services").(string)
+	cli, ok := p.Has("client")
+	if !ok {
+		return nil, errors.New("No etcd client found.")
+	}
+	client := cli.(Watcher)
+
+	// Stupid hack because etcd watch seems to be broken, constantly complaining
+	// that the JSON it received is malformed.
+	safely.GoDo(c, func() {
+		for {
+			response, err := client.Watch(path, 0, true, nil, nil)
+			if err != nil {
+				log.Errf(c, "Etcd Watch failed: %s", err)
+				time.Sleep(50 * time.Millisecond)
+				continue
+			}
+
+			if response.Node == nil {
+				log.Infof(c, "Unexpected Etcd message: %v", response)
+			}
+			git := exec.Command("/home/git/check-repos")
+			if out, err := git.CombinedOutput(); err != nil {
+				log.Errf(c, "Failed git check-repos: %s", err)
+				log.Infof(c, "Output: %s", out)
+			}
+		}
+
+	})
+
+	return nil, nil
+
+	/* Watch seems to be broken. So we do this stupid watch loop instead.
+	receiver := make(chan *etcd.Response)
+	stop := make(chan bool)
+	// Buffer the channels so that we don't hang waiting for go-etcd to
+	// read off the channel.
+	stopetcd := make(chan bool, 1)
+	stopwatch := make(chan bool, 1)
+
+
+	// Watch for errors.
+	safely.GoDo(c, func() {
+		// When a receiver is passed in, no *Response is ever returned. Instead,
+		// Watch acts like an error channel, and receiver gets all of the messages.
+		_, err := client.Watch(path, 0, true, receiver, stopetcd)
+		if err != nil {
+			log.Infof(c, "Watcher stopped with error '%s'", err)
+			stopwatch <- true
+			//close(stopwatch)
+		}
+	})
+	// Watch for events
+	safely.GoDo(c, func() {
+		for {
+			select {
+			case msg := <-receiver:
+				if msg.Node != nil {
+					log.Infof(c, "Received notification %s for %s", msg.Action, msg.Node.Key)
+				} else {
+					log.Infof(c, "Received unexpected etcd message: %v", msg)
+				}
+				git := exec.Command("/home/git/check-repos")
+				if out, err := git.CombinedOutput(); err != nil {
+					log.Errf(c, "Failed git check-repos: %s", err)
+					log.Infof(c, "Output: %s", out)
+				}
+			case <-stopwatch:
+				c.Logf("debug", "Received signal to stop watching events.")
+				return
+			}
+		}
+	})
+	// Fan out stop requests.
+	safely.GoDo(c, func() {
+		<-stop
+		stopwatch <- true
+		stopetcd <- true
+		close(stopwatch)
+		close(stopetcd)
+	})
+
+	return stop, nil
+	*/
+}
+
+// checkRetry overrides etcd.DefaultCheckRetry.
+//
+// It adds configurable number of retries and configurable timesouts.
+func checkRetry(c *etcd.Cluster, numReqs int, last http.Response, err error) error {
+
+	if numReqs > retryCycles*len(c.Machines) {
+		return fmt.Errorf("Tried and failed %d cluster connections: %s", retryCycles, err)
+	}
+
+	switch last.StatusCode {
+	case 0:
+		return nil
+	case 500:
```

Comment:
That's a good question. Since there's no code 0 constant, is it more readable to use the long constant names and leave 0, or keep them all ints? I prefer for all of the cases to look the same because it requires less thinking.